src/hotspot/share/gc/shenandoah/shenandoahFreeSet.hpp

```diff
@@ -28,69 +29,489 @@
 #include "gc/shenandoah/shenandoahHeapRegionSet.hpp"
 #include "gc/shenandoah/shenandoahHeap.hpp"
 
+// The API and internal implementation of ShenandoahSimpleBitMap and ShenandoahRegionPartitions use ssize_t to
+// represent index, even though index is "inherently" unsigned.  There are two reasons for this choice:
+//  1. We use -1 as a sentinel value to represent empty partitions.  This same value may be used to represent
+//     failure to find a previous set bit or previous range of set bits.
+//  2. Certain loops are written most naturally if the iterator, which may hold the sentinel -1 value, can be
+//     declared as signed and the terminating condition can be < 0.
+
+
+
+// ShenandoahSimpleBitMap resembles CHeapBitMap but adds missing support for find_next_consecutive_bits() and
+// find_prev_contiguous_bits.  An alternative refactoring of code would subclass CHeapBitMap, but this might
+// break abstraction rules, because efficient implementation requires assumptions about superclass internals that
+// might be violatee through future software maintenance.
+class ShenandoahSimpleBitMap {
```

Comment:
I think this class should go into its own set of files. It would certainly help readability of ShFreeSet which is now somewhat dominated by ShSimpleBitMap.

Reply:
Thanks for suggestion.  Done.

Reply:
Thanks.  I've separated this code into distinct source files.